src/lib.rs

```diff
@@ -676,6 +789,35 @@ impl<Types: SwarmTypes> Future for IpfsFuture<Types> {
                         let wantlist = self.swarm.bitswap().local_wantlist();
                         let _ = ret.send((stats, peers, wantlist).into());
                     }
+                    IpfsEvent::AddListeningAddress(addr, ret) => {
+                        if self.listening_addresses.contains_key(&addr) {
+                            let _ = ret.send(Err(format_err!("Already adding a possibly ephemeral multiaddr, wait first one to resolve before adding next: {}", addr)));
+                        } else {
+                            match Swarm::listen_on(&mut self.swarm, addr.clone()) {
+                                Ok(id) => {
+                                    self.listening_addresses.insert(addr, (id, Some(ret)));
+                                }
+                                Err(e) => {
+                                    let _ = ret.send(Err(Error::from(e)));
+                                }
+                            }
+                        }
+                    }
```

Comment:
Clippy would prefer using `self.listening_addresses.entry(addr)` but this would involve having two mutable references through `Pin<&mut Self>` which at least didn't succeed on the obvious try:

```
match self.listening_addresses.entry(addr) {
  Entry::Occupied(oe) => // send err, use oe.key() for error msg
  Entry::Vacant(ve) => {
    match Swarm::listen_on(&mut self.swarm, ve.key().to_owned()) { // ERROR: second borrow
    }
  }
}```